File: modules/edumfa-rsa/android/src/main/java/expo/modules/edumfarsa/EduMfaRsaModule.kt
package expo.modules.edumfarsa

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.PrivateKey
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec

private const val ANDROID_KEYSTORE = "AndroidKeyStore"
private const val RSA_ALGORITHM = KeyProperties.KEY_ALGORITHM_RSA
private const val SHA256_WITH_RSA = "SHA256withRSA"
private const val SHA512_WITH_RSA = "SHA512withRSA"
private const val SHA1_WITH_RSA = "SHA1withRSA"
private const val SHA256 = "SHA256"
private const val SHA512 = "SHA512"
private const val SHA1 = "SHA1"

private val RSA_ALGORITHM_IDENTIFIER = byteArrayOf(
  0x30, 0x0d, 0x06, 0x09, 0x2a, 0x86.toByte(), 0x48, 0x86.toByte(),
  0xf7.toByte(), 0x0d, 0x01, 0x01, 0x01, 0x05, 0x00
)

class EduMfaRsaModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("EduMfaRsa")

    Constant("SHA256") { SHA256 }
    Constant("SHA512") { SHA512 }
    Constant("SHA1") { SHA1 }

    AsyncFunction("generateKeyPair") { keyAlias: String, keySize: Int ->
      mapOf("publicKey" to generateKeyPair(keyAlias, keySize))
    }

    AsyncFunction("getPublicKey") { keyAlias: String ->
      mapOf("publicKey" to getPublicKeyPem(keyAlias))
    }

    AsyncFunction("sign") { message: String, keyAlias: String, algorithm: String ->
      sign(message, keyAlias, algorithm)
    }

    AsyncFunction("verify") { message: String, signatureBase64: String, publicKey: String, algorithm: String ->
      verify(message, signatureBase64, publicKey, algorithm)
    }

    AsyncFunction("deleteKeyPair") { keyAlias: String ->
      deleteKeyPair(keyAlias)
    }
  }

  private fun generateKeyPair(keyAlias: String, keySize: Int): String {
    validateKeyAlias(keyAlias)
    validateKeySize(keySize)

    val keyStore = getKeyStore()
    if (keyStore.containsAlias(keyAlias)) {
      keyStore.deleteEntry(keyAlias)
    }

    val generator = KeyPairGenerator.getInstance(RSA_ALGORITHM, ANDROID_KEYSTORE)
    val spec = KeyGenParameterSpec.Builder(
      keyAlias,
      KeyProperties.PURPOSE_SIGN or KeyProperties.PURPOSE_VERIFY
    )
      .setKeySize(keySize)
      .setDigests(
        KeyProperties.DIGEST_SHA256,
        KeyProperties.DIGEST_SHA512,
        KeyProperties.DIGEST_SHA1
      )
      .setSignaturePaddings(KeyProperties.SIGNATURE_PADDING_RSA_PKCS1)
      .setUserAuthenticationRequired(false)
      .build()

    generator.initialize(spec)
    val keyPair = generator.generateKeyPair()
    return pem("PUBLIC KEY", keyPair.public.encoded)
  }

  private fun getPublicKeyPem(keyAlias: String): String {
    validateKeyAlias(keyAlias)
    return pem("PUBLIC KEY", getPrivateKeyEntry(keyAlias).certificate.publicKey.encoded)
  }

  private fun sign(message: String, keyAlias: String, algorithm: String): String {
    validateKeyAlias(keyAlias)

    val privateKey = getPrivateKeyEntry(keyAlias).privateKey
    val signature = Signature.getInstance(signatureAlgorithm(algorithm))
    signature.initSign(privateKey)
    signature.update(message.toByteArray(Charsets.UTF_8))
    return Base64.encodeToString(signature.sign(), Base64.NO_WRAP)
  }

  private fun verify(
    message: String,
    signatureBase64: String,
    publicKey: String,
    algorithm: String
  ): Boolean {
    val signatureBytes = Base64.decode(signatureBase64, Base64.DEFAULT)
    val verifier = Signature.getInstance(signatureAlgorithm(algorithm))
    verifier.initVerify(parsePublicKey(publicKey))
    verifier.update(message.toByteArray(Charsets.UTF_8))
    return verifier.verify(signatureBytes)
  }

  private fun deleteKeyPair(keyAlias: String): Boolean {
    validateKeyAlias(keyAlias)
    val keyStore = getKeyStore()
    if (keyStore.containsAlias(keyAlias)) {
      keyStore.deleteEntry(keyAlias)
    }
    return true
  }

  private fun getPrivateKeyEntry(keyAlias: String): KeyStore.PrivateKeyEntry {
    val keyStore = getKeyStore()
    val entry = keyStore.getEntry(keyAlias, null)
    if (entry !is KeyStore.PrivateKeyEntry) {
      throw IllegalStateException("No RSA private key found for alias '$keyAlias'.")
    }

    val privateKey: PrivateKey = entry.privateKey
    if (privateKey.algorithm != RSA_ALGORITHM) {
      throw IllegalStateException("Key '$keyAlias' is not an RSA private key.")
    }

    return entry
  }

  private fun getKeyStore(): KeyStore {
    return KeyStore.getInstance(ANDROID_KEYSTORE).apply {
      load(null)
    }
  }

  private fun parsePublicKey(publicKey: String): PublicKey {
    val der = decodePemBody(publicKey)
    val keyFactory = KeyFactory.getInstance(RSA_ALGORITHM)

    return try {
      keyFactory.generatePublic(X509EncodedKeySpec(der))
    } catch (_: Exception) {
      keyFactory.generatePublic(X509EncodedKeySpec(wrapPkcs1PublicKeyInSubjectPublicKeyInfo(der)))
    }
  }

  private fun decodePemBody(key: String): ByteArray {
    var body = key
      .replace(Regex("-----BEGIN [^-]+-----"), "")
      .replace(Regex("-----END [^-]+-----"), "")
      .replace(Regex("\\s"), "")
      .replace('-', '+')
      .replace('_', '/')

    if (body.isBlank()) {
      throw IllegalArgumentException("The RSA public key is empty.")
    }

    val paddingLength = (4 - (body.length % 4)) % 4
    if (paddingLength > 0) {
      body += "=".repeat(paddingLength)
    }

    return Base64.decode(body, Base64.DEFAULT)
  }

  private fun wrapPkcs1PublicKeyInSubjectPublicKeyInfo(pkcs1: ByteArray): ByteArray {
    val bitString = byteArrayOf(0x03) +
      derLength(pkcs1.size + 1) +
      byteArrayOf(0x00) +
      pkcs1
    val body = RSA_ALGORITHM_IDENTIFIER + bitString
    return byteArrayOf(0x30) + derLength(body.size) + body
  }

  private fun derLength(length: Int): ByteArray {
    if (length < 128) {
      return byteArrayOf(length.toByte())
    }

    var value = length
    val bytes = mutableListOf<Byte>()
    while (value > 0) {
      bytes.add(0, (value and 0xff).toByte())
      value = value ushr 8
    }

    return byteArrayOf((0x80 or bytes.size).toByte()) + bytes.toByteArray()
  }

  private fun pem(label: String, data: ByteArray): String {
    val base64 = Base64.encodeToString(data, Base64.NO_WRAP)
    return "-----BEGIN $label-----\n${base64.chunked(64).joinToString("\n")}\n-----END $label-----"
  }

  private fun validateKeyAlias(keyAlias: String) {
    require(keyAlias.isNotBlank()) {
      "Key alias must not be empty."
    }
  }

  private fun validateKeySize(keySize: Int) {
    require(keySize >= 2048 && keySize % 1024 == 0) {
      "Invalid RSA key size: $keySize."
    }
  }

  private fun signatureAlgorithm(algorithm: String): String {
    return when (algorithm) {
      SHA256 -> SHA256_WITH_RSA
      SHA512 -> SHA512_WITH_RSA
      SHA1 -> SHA1_WITH_RSA
      else -> throw IllegalArgumentException("Unsupported RSA signature algorithm: $algorithm.")
    }
  }
}

File: modules/edumfa-rsa/ios/EduMfaRsaModule.swift
import ExpoModulesCore
import Foundation
import Security

private enum EduMfaRsaError: Error, LocalizedError {
  case emptyAlias
  case invalidKeySize(Int)
  case keyGenerationFailed(String)
  case keyNotFound(String)
  case publicKeyExportFailed
  case invalidPublicKey
  case invalidSignature
  case unsupportedAlgorithm(String)
  case signingFailed(String)
  case keychainDeleteFailed(OSStatus)

  var errorDescription: String? {
    switch self {
    case .emptyAlias:
      return "Key alias must not be empty."
    case .invalidKeySize(let keySize):
      return "Invalid RSA key size: \(keySize)."
    case .keyGenerationFailed(let reason):
      return "RSA key generation failed: \(reason)"
    case .keyNotFound(let keyAlias):
      return "No RSA private key found for alias '\(keyAlias)'."
    case .publicKeyExportFailed:
      return "Could not export the RSA public key."
    case .invalidPublicKey:
      return "The RSA public key is not valid."
    case .invalidSignature:
      return "The RSA signature is not valid base64."
    case .unsupportedAlgorithm(let algorithm):
      return "Unsupported RSA signature algorithm: \(algorithm)."
    case .signingFailed(let reason):
      return "RSA signing failed: \(reason)"
    case .keychainDeleteFailed(let status):
      return "Could not delete RSA keychain item. OSStatus: \(status)."
    }
  }
}

private struct Asn1Reader {
  private let bytes: [UInt8]
  private var offset = 0

  init(_ data: Data) {
    self.bytes = Array(data)
  }

  init(_ bytes: [UInt8]) {
    self.bytes = bytes
  }

  mutating func peekTag() throws -> UInt8 {
    guard offset < bytes.count else {
      throw EduMfaRsaError.invalidPublicKey
    }
    return bytes[offset]
  }

  mutating func readElement(expectedTag: UInt8) throws -> [UInt8] {
    guard offset < bytes.count else {
      throw EduMfaRsaError.invalidPublicKey
    }

    let tag = bytes[offset]
    offset += 1
    guard tag == expectedTag else {
      throw EduMfaRsaError.invalidPublicKey
    }

    let length = try readLength()
    guard offset + length <= bytes.count else {
      throw EduMfaRsaError.invalidPublicKey
    }

    let content = Array(bytes[offset..<(offset + length)])
    offset += length
    return content
  }

  private mutating func readLength() throws -> Int {
    guard offset < bytes.count else {
      throw EduMfaRsaError.invalidPublicKey
    }

    let first = bytes[offset]
    offset += 1

    if first & 0x80 == 0 {
      return Int(first)
    }

    let lengthByteCount = Int(first & 0x7f)
    guard lengthByteCount > 0, lengthByteCount <= 4, offset + lengthByteCount <= bytes.count else {
      throw EduMfaRsaError.invalidPublicKey
    }

    var length = 0
    for _ in 0..<lengthByteCount {
      length = (length << 8) | Int(bytes[offset])
      offset += 1
    }
    return length
  }
}

public class EduMfaRsaModule: Module {
  public func definition() -> ModuleDefinition {
    Name("EduMfaRsa")

    Constant("SHA256") { "SHA256" }
    Constant("SHA512") { "SHA512" }
    Constant("SHA1") { "SHA1" }

    AsyncFunction("generateKeyPair") { (keyAlias: String, keySize: Int) -> [String: String] in
      let publicKey = try self.generateKeyPair(keyAlias: keyAlias, keySize: keySize)
      return ["publicKey": publicKey]
    }

    AsyncFunction("getPublicKey") { (keyAlias: String) -> [String: String] in
      let publicKey = try self.getPublicKeyPem(keyAlias: keyAlias)
      return ["publicKey": publicKey]
    }

    AsyncFunction("sign") { (message: String, keyAlias: String, algorithm: String) -> String in
      return try self.sign(message: message, keyAlias: keyAlias, algorithm: algorithm)
    }

    AsyncFunction("verify") { (message: String, signatureBase64: String, publicKey: String, algorithm: String) -> Bool in
      return try self.verify(message: message, signatureBase64: signatureBase64, publicKey: publicKey, algorithm: algorithm)
    }

    AsyncFunction("deleteKeyPair") { (keyAlias: String) -> Bool in
      return try self.deleteKeyPair(keyAlias: keyAlias)
    }
  }

  private func generateKeyPair(keyAlias: String, keySize: Int) throws -> String {
    try validate(keyAlias: keyAlias, keySize: keySize)
    try deleteKeychainTag(privateKeyTag(for: keyAlias), allowNotFound: true)
    try deleteKeychainTag(publicKeyTag(for: keyAlias), allowNotFound: true)

    let attributes: [String: Any] = [
      String(kSecAttrKeyType): kSecAttrKeyTypeRSA,
      String(kSecAttrKeySizeInBits): keySize,
      String(kSecPrivateKeyAttrs): [
        String(kSecAttrIsPermanent): true,
        String(kSecAttrApplicationTag): privateKeyTag(for: keyAlias),
        String(kSecAttrAccessible): kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
      ]
    ]

    var error: Unmanaged<CFError>?
    guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
      let reason = error?.takeRetainedValue().localizedDescription ?? "unknown error"
      throw EduMfaRsaError.keyGenerationFailed(reason)
    }

    guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
      throw EduMfaRsaError.publicKeyExportFailed
    }

    return try publicKeyPem(from: publicKey)
  }

  private func getPublicKeyPem(keyAlias: String) throws -> String {
    try validate(keyAlias: keyAlias)
    let privateKey = try getPrivateKey(keyAlias: keyAlias)
    guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
      throw EduMfaRsaError.publicKeyExportFailed
    }
    return try publicKeyPem(from: publicKey)
  }

  private func sign(message: String, keyAlias: String, algorithm: String) throws -> String {
    try validate(keyAlias: keyAlias)
    let privateKey = try getPrivateKey(keyAlias: keyAlias)
    let secAlgorithm = try signatureAlgorithm(algorithm)

    guard SecKeyIsAlgorithmSupported(privateKey, .sign, secAlgorithm) else {
      throw EduMfaRsaError.unsupportedAlgorithm(algorithm)
    }

    let messageData = Data(message.utf8)
    var error: Unmanaged<CFError>?
    guard let signature = SecKeyCreateSignature(privateKey, secAlgorithm, messageData as CFData, &error) as Data? else {
      let reason = error?.takeRetainedValue().localizedDescription ?? "unknown error"
      throw EduMfaRsaError.signingFailed(reason)
    }

    return signature.base64EncodedString()
  }

  private func verify(message: String, signatureBase64: String, publicKey: String, algorithm: String) throws -> Bool {
    let key = try createPublicKey(from: publicKey)
    let secAlgorithm = try signatureAlgorithm(algorithm)

    guard SecKeyIsAlgorithmSupported(key, .verify, secAlgorithm) else {
      throw EduMfaRsaError.unsupportedAlgorithm(algorithm)
    }

    guard let signature = Data(base64Encoded: signatureBase64, options: .ignoreUnknownCharacters) else {
      throw EduMfaRsaError.invalidSignature
    }

    let messageData = Data(message.utf8)
    var error: Unmanaged<CFError>?
    let isValid = SecKeyVerifySignature(key, secAlgorithm, messageData as CFData, signature as CFData, &error)
    if !isValid {
      _ = error?.takeRetainedValue()
    }
    return isValid
  }

  private func deleteKeyPair(keyAlias: String) throws -> Bool {
    try validate(keyAlias: keyAlias)
    try deleteKeychainTag(privateKeyTag(for: keyAlias), allowNotFound: true)
    try deleteKeychainTag(publicKeyTag(for: keyAlias), allowNotFound: true)
    return true
  }

  private func getPrivateKey(keyAlias: String) throws -> SecKey {
    let query: [String: Any] = [
      String(kSecClass): kSecClassKey,
      String(kSecAttrKeyType): kSecAttrKeyTypeRSA,
      String(kSecAttrKeyClass): kSecAttrKeyClassPrivate,
      String(kSecAttrApplicationTag): privateKeyTag(for: keyAlias),
      String(kSecReturnRef): true
    ]

    var item: CFTypeRef?
    let status = SecItemCopyMatching(query as CFDictionary, &item)
    guard status == errSecSuccess, let privateKey = item else {
      throw EduMfaRsaError.keyNotFound(keyAlias)
    }
    return privateKey as! SecKey
  }

  private func deleteKeychainTag(_ tag: Data, allowNotFound: Bool) throws {
    let query: [String: Any] = [
      String(kSecClass): kSecClassKey,
      String(kSecAttrKeyType): kSecAttrKeyTypeRSA,
      String(kSecAttrApplicationTag): tag
    ]

    let status = SecItemDelete(query as CFDictionary)
    if status != errSecSuccess && !(allowNotFound && status == errSecItemNotFound) {
      throw EduMfaRsaError.keychainDeleteFailed(status)
    }
  }

  private func publicKeyPem(from publicKey: SecKey) throws -> String {
    var error: Unmanaged<CFError>?
    guard let pkcs1 = SecKeyCopyExternalRepresentation(publicKey, &error) as Data? else {
      _ = error?.takeRetainedValue()
      throw EduMfaRsaError.publicKeyExportFailed
    }

    let spki = wrapPkcs1PublicKeyInSubjectPublicKeyInfo(pkcs1)
    return pem(label: "PUBLIC KEY", data: spki)
  }

  private func createPublicKey(from publicKey: String) throws -> SecKey {
    let der = try decodePemBody(publicKey)
    let attributes: [String: Any] = [
      String(kSecAttrKeyType): kSecAttrKeyTypeRSA,
      String(kSecAttrKeyClass): kSecAttrKeyClassPublic
    ]

    var error: Unmanaged<CFError>?
    if let key = SecKeyCreateWithData(der as CFData, attributes as CFDictionary, &error) {
      return key
    }
    _ = error?.takeRetainedValue()

    let pkcs1 = pkcs1PublicKeyData(fromPossibleSubjectPublicKeyInfo: der)
    error = nil
    if let key = SecKeyCreateWithData(pkcs1 as CFData, attributes as CFDictionary, &error) {
      return key
    }
    _ = error?.takeRetainedValue()

    throw EduMfaRsaError.invalidPublicKey
  }

  private func decodePemBody(_ key: String) throws -> Data {
    var body = key
      .replacingOccurrences(of: "-----BEGIN [^-]+-----", with: "", options: .regularExpression)
      .replacingOccurrences(of: "-----END [^-]+-----", with: "", options: .regularExpression)
      .replacingOccurrences(of: "\\s", with: "", options: .regularExpression)
      .replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")

    let paddingLength = (4 - (body.count % 4)) % 4
    if paddingLength > 0 {
      body.append(String(repeating: "=", count: paddingLength))
    }

    guard let data = Data(base64Encoded: body, options: .ignoreUnknownCharacters) else {
      throw EduMfaRsaError.invalidPublicKey
    }
    return data
  }

  private func pkcs1PublicKeyData(fromPossibleSubjectPublicKeyInfo data: Data) -> Data {
    do {
      var reader = Asn1Reader(data)
      let outerContent = try reader.readElement(expectedTag: 0x30)
      var outer = Asn1Reader(outerContent)
      guard try outer.peekTag() == 0x30 else {
        return data
      }
      _ = try outer.readElement(expectedTag: 0x30)
      let bitString = try outer.readElement(expectedTag: 0x03)
      guard bitString.first == 0x00 else {
        return data
      }
      return Data(bitString.dropFirst())
    } catch {
      return data
    }
  }

  private func wrapPkcs1PublicKeyInSubjectPublicKeyInfo(_ pkcs1: Data) -> Data {
    let algorithmIdentifier: [UInt8] = [
      0x30, 0x0d, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86,
      0xf7, 0x0d, 0x01, 0x01, 0x01, 0x05, 0x00
    ]
    let bitString = Data([UInt8(0x03)] + derLength(pkcs1.count + 1) + [UInt8(0x00)] + Array(pkcs1))
    var body = Data(algorithmIdentifier)
    body.append(bitString)
    return Data([UInt8(0x30)] + derLength(body.count) + Array(body))
  }

  private func derLength(_ length: Int) -> [UInt8] {
    if length < 128 {
      return [UInt8(length)]
    }

    var value = length
    var bytes: [UInt8] = []
    while value > 0 {
      bytes.insert(UInt8(value & 0xff), at: 0)
      value >>= 8
    }
    return [UInt8(0x80) | UInt8(bytes.count)] + bytes
  }

  private func pem(label: String, data: Data) -> String {
    let base64 = data.base64EncodedString()
    let lines = stride(from: 0, to: base64.count, by: 64).map { index -> String in
      let start = base64.index(base64.startIndex, offsetBy: index)
      let end = base64.index(start, offsetBy: min(64, base64.distance(from: start, to: base64.endIndex)))
      return String(base64[start..<end])
    }
    return "-----BEGIN \(label)-----\n\(lines.joined(separator: "\n"))\n-----END \(label)-----"
  }

  private func signatureAlgorithm(_ algorithm: String) throws -> SecKeyAlgorithm {
    switch algorithm {
    case "SHA256":
      return .rsaSignatureMessagePKCS1v15SHA256
    case "SHA512":
      return .rsaSignatureMessagePKCS1v15SHA512
    case "SHA1":
      return .rsaSignatureMessagePKCS1v15SHA1
    default:
      throw EduMfaRsaError.unsupportedAlgorithm(algorithm)
    }
  }

  private func validate(keyAlias: String, keySize: Int? = nil) throws {
    guard !keyAlias.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      throw EduMfaRsaError.emptyAlias
    }

    if let keySize, keySize < 2048 || keySize % 1024 != 0 {
      throw EduMfaRsaError.invalidKeySize(keySize)
    }
  }

  private func privateKeyTag(for keyAlias: String) -> Data {
    Data("\(keyAlias).private".utf8)
  }

  private func publicKeyTag(for keyAlias: String) -> Data {
    Data("\(keyAlias).public".utf8)
  }
}
